config.yml

```diff
@@ -1,98 +1,121 @@
+#
+#  * Delete the `delete: true` sections of the config after at least successful run of GitLabForm over the appropriate
+#    projects or groups. This will make GitLabForm will run faster as it won't try to remove those things over and over.
 
-common_settings:
-  # common settings for ALL projects in ALL groups
-
-  # keys and values here are as described at https://docs.gitlab.com/ee/api/project_level_variables.html#create-variable
-  secret_variables:
-    # this name is not actually used, it's just for you
-    a_secret_you_want_to_add_to_all_projects_in_your_gitlab_instance:
-      key: A_VERY_COMMONLY_USED_PASSWORD
-      value: "ThisIsAVerySecretPassword"
-      delete: false  # Set to true to delete this variable
-
-  # keys and values here are as described at https://docs.gitlab.com/ee/api/group_level_variables.html#create-variable
-  group_secret_variables:
-    # this name is not actually used, it's just for you
-    a_secret_you_want_to_add_to_all_groups_in_your_gitlab_instance:
-      key: A_VERY_COMMONLY_USED_PASSWORD
-      value: "ThisIsAVerySecretPassword"
-      variable_type: env_var #or file
-      protected: false #
-      delete: false  # Set to true to delete this variable
-
-group_settings:
-  # settings for ALL projects in 'my-group' group
-  "my-group":
-    deploy_keys:
-      # this name is not actually used, it's just for you
-      a_friendly_deploy_key_name:
-        # you have to pass whole SSH key content here even if GitLab already has this key added and you just assign it
-        # to another project here
-        # this is a limitation of GitLab API
-        key: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDB2QKx6BPzL (...)"
-        title: ssh_key_name_that_is_shown_in_gitlab
-        # note that you can set this to `true` or `false` only on the first assignment of the key / its creation
-        # see https://gitlab.com/gitlab-org/gitlab-ce/issues/30021#note_39567845
-        # this is a limitation of GitLab API
-        can_push: false
-      # this name is not actually used, it's just for you
-      another_friendly_deploy_key_name:
-        key: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDtbyEK66RXg (...)"
-        title: other_ssh_key_name_that_is_shown_in_gitlab
-        can_push: false
 
-    # keys and values here are as described at https://docs.gitlab.com/ee/api/project_level_variables.html#create-variable
-    secret_variables:
-      # this name is not actually used, it's just for you
-      a_friendly_secret_variable_name:
-        key: SSH_PRIVATE_KEY_BASE64
-        value: "LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUl (...)"
+# The main config starts with this key
+projects_and_groups:
+
+  # On this indentation level you choose what groups and/or projects you want to configure.
+  #
+  # Here are examples of possible dict key.
+  #
+  # This matches all the groups AND all the projects EXCEPT personal projects!
+  # "*":
```

Comment:
Pattern matching would make the syntax very powerful indeed. The one potential issue I see here is: what if there are 2+ different patterns matching the same project or same group?

* would this be considered a "fatal error"?
* if not, which pattern wins?

Reply:
Thanks for pointing this issue out. We haven't made a decision yet, but we will define some behaviour and write tests to ensure that it will be stable.
And what would be your suggestion?

Reply:
I actually quite liked the previous 3-level priority: common, overridden on group level, overridden on project level. 
This was simple and IMO very effective.

To make it work with pattern matching, I suppose something like this should work:

```yaml
projects_and_groups:
  - path: '*'   # equivalent of "common_settings"
    group_settings:
      ...
    project_settings:
  - path: 'foogroup/*'   # equivalent of group settings (although it doesn't make it clear whether this includes sub-groups or not)
    group_settings:
      ..
  - path: 'bargroup/quxproject'  # and this is an override per project
    project_settings:
      ..
```

Those could be processed from top to bottom, with the settings at the bottom "winning" to allow to merge them.

A more powerful option (but also even more complicated):

```yaml
projects_and_groups:
  - path: '*'   # equivalent of "common_settings"
    continue: true #### <<---- let user select whether this rule should end processing or not
    group_settings:
      ...
    project_settings:
  - path: 'foogroup/*'   # equivalent of group settings (although it doesn't make it clear whether this includes sub-groups or not)
    continue: false #### <<---- let user select whether this rule should end processing or not
    group_settings:
      ..
  - path: 'foogroup/quxproject'  # and this is an override per project 
    project_settings:
      ..
```
```

Reply:
That's a nice proposal! I believe with tree-like structure of GitLab groups and projects, this could cover most (if not all) of the cases - especially assuming that patterns are defined in the topological order (with patterns affecting more top-level groups/projects defined before the other patterns)

Reply:
Wondering about pattern matching: unless resorting to "regex", it could be helpful to intrduce a distinction between `/*` and `/**` patterns where the former would mean only direct descendents of a group and the latter a recursive match.

Rationale: if recursive were to be default, it would be quite difficult to partition the groups sensibly (make any form of exclusion).

Reply:
Hey @weakcamel! I think that it would be more efficient to discuss and come up with the best solution using our new Slack workspace. If you (or anyone else reading this and interested in it) want to join it the please use this invite link: https://join.slack.com/t/gitlabform/shared_invite/zt-ljgzvxk0-TubjJFA5lLpRkyivQWI_Qg (valid until ~21 Feb).